README.markdown
# EXAMPLE

Like any Deferrable eventmachine-based protocol implementation, using EM-Couchbase involves making calls and passing blocks that serve as callbacks when the call returns.

    require 'em-couchbase'

    EM.run do
      couchbase = EM::Protocols::Couchbase.connect
      couchbase.set "a", "foo" do |response|
        if response.success?
          couchbase.get "a" do |response|
            puts response.inspect
          end
        end
      end
    end


HISTORY.markdown
## 0.2.0 / 2012-07-24

* Support INCR and DECR commands
* Change ruby-debug dependency to debugger

## 0.1.0 / 2012-03-27

* Client bootstrap
* Handles cluster upgrades (adding/removing nodes)
* Support GET and SET commands


lib/em-couchbase/util.rb
#/* The crc32 functions and data was originally written by Spencer
# * Garrett <srg@quick.com> and was gleaned from the PostgreSQL source
# * tree via the files contrib/ltree/crc32.[ch] and from FreeBSD at
# * src/usr.bin/cksum/crc32.c.
# */


module EventMachine
  module Protocols

    module Couchbase

      module Util

        CRC32TAB = [
          0x00000000, 0x77073096, 0xee0e612c, 0x990951ba,
          0x076dc419, 0x706af48f, 0xe963a535, 0x9e6495a3,
          0x0edb8832, 0x79dcb8a4, 0xe0d5e91e, 0x97d2d988,
          0x09b64c2b, 0x7eb17cbd, 0xe7b82d07, 0x90bf1d91,
          0x1db71064, 0x6ab020f2, 0xf3b97148, 0x84be41de,
          0x1adad47d, 0x6ddde4eb, 0xf4d4b551, 0x83d385c7,
          0x136c9856, 0x646ba8c0, 0xfd62f97a, 0x8a65c9ec,
          0x14015c4f, 0x63066cd9, 0xfa0f3d63, 0x8d080df5,
          0x3b6e20c8, 0x4c69105e, 0xd56041e4, 0xa2677172,
          0x3c03e4d1, 0x4b04d447, 0xd20d85fd, 0xa50ab56b,
          0x35b5a8fa, 0x42b2986c, 0xdbbbc9d6, 0xacbcf940,
          0x32d86ce3, 0x45df5c75, 0xdcd60dcf, 0xabd13d59,
          0x26d930ac, 0x51de003a, 0xc8d75180, 0xbfd06116,
          0x21b4f4b5, 0x56b3c423, 0xcfba9599, 0xb8bda50f,
          0x2802b89e, 0x5f058808, 0xc60cd9b2, 0xb10be924,
          0x2f6f7c87, 0x58684c11, 0xc1611dab, 0xb6662d3d,
          0x76dc4190, 0x01db7106, 0x98d220bc, 0xefd5102a,
          0x71b18589, 0x06b6b51f, 0x9fbfe4a5, 0xe8b8d433,
          0x7807c9a2, 0x0f00f934, 0x9609a88e, 0xe10e9818,
          0x7f6a0dbb, 0x086d3d2d, 0x91646c97, 0xe6635c01,
          0x6b6b51f4, 0x1c6c6162, 0x856530d8, 0xf262004e,
          0x6c0695ed, 0x1b01a57b, 0x8208f4c1, 0xf50fc457,
          0x65b0d9c6, 0x12b7e950, 0x8bbeb8ea, 0xfcb9887c,
          0x62dd1ddf, 0x15da2d49, 0x8cd37cf3, 0xfbd44c65,
          0x4db26158, 0x3ab551ce, 0xa3bc0074, 0xd4bb30e2,
          0x4adfa541, 0x3dd895d7, 0xa4d1c46d, 0xd3d6f4fb,
          0x4369e96a, 0x346ed9fc, 0xad678846, 0xda60b8d0,
          0x44042d73, 0x33031de5, 0xaa0a4c5f, 0xdd0d7cc9,
          0x5005713c, 0x270241aa, 0xbe0b1010, 0xc90c2086,
          0x5768b525, 0x206f85b3, 0xb966d409, 0xce61e49f,
          0x5edef90e, 0x29d9c998, 0xb0d09822, 0xc7d7a8b4,
          0x59b33d17, 0x2eb40d81, 0xb7bd5c3b, 0xc0ba6cad,
          0xedb88320, 0x9abfb3b6, 0x03b6e20c, 0x74b1d29a,
          0xead54739, 0x9dd277af, 0x04db2615, 0x73dc1683,
          0xe3630b12, 0x94643b84, 0x0d6d6a3e, 0x7a6a5aa8,
          0xe40ecf0b, 0x9309ff9d, 0x0a00ae27, 0x7d079eb1,
          0xf00f9344, 0x8708a3d2, 0x1e01f268, 0x6906c2fe,
          0xf762575d, 0x806567cb, 0x196c3671, 0x6e6b06e7,
          0xfed41b76, 0x89d32be0, 0x10da7a5a, 0x67dd4acc,
          0xf9b9df6f, 0x8ebeeff9, 0x17b7be43, 0x60b08ed5,
          0xd6d6a3e8, 0xa1d1937e, 0x38d8c2c4, 0x4fdff252,
          0xd1bb67f1, 0xa6bc5767, 0x3fb506dd, 0x48b2364b,
          0xd80d2bda, 0xaf0a1b4c, 0x36034af6, 0x41047a60,
          0xdf60efc3, 0xa867df55, 0x316e8eef, 0x4669be79,
          0xcb61b38c, 0xbc66831a, 0x256fd2a0, 0x5268e236,
          0xcc0c7795, 0xbb0b4703, 0x220216b9, 0x5505262f,
          0xc5ba3bbe, 0xb2bd0b28, 0x2bb45a92, 0x5cb36a04,
          0xc2d7ffa7, 0xb5d0cf31, 0x2cd99e8b, 0x5bdeae1d,
          0x9b64c2b0, 0xec63f226, 0x756aa39c, 0x026d930a,
          0x9c0906a9, 0xeb0e363f, 0x72076785, 0x05005713,
          0x95bf4a82, 0xe2b87a14, 0x7bb12bae, 0x0cb61b38,
          0x92d28e9b, 0xe5d5be0d, 0x7cdcefb7, 0x0bdbdf21,
          0x86d3d2d4, 0xf1d4e242, 0x68ddb3f8, 0x1fda836e,
          0x81be16cd, 0xf6b9265b, 0x6fb077e1, 0x18b74777,
          0x88085ae6, 0xff0f6a70, 0x66063bca, 0x11010b5c,
          0x8f659eff, 0xf862ae69, 0x616bffd3, 0x166ccf45,
          0xa00ae278, 0xd70dd2ee, 0x4e048354, 0x3903b3c2,
          0xa7672661, 0xd06016f7, 0x4969474d, 0x3e6e77db,
          0xaed16a4a, 0xd9d65adc, 0x40df0b66, 0x37d83bf0,
          0xa9bcae53, 0xdebb9ec5, 0x47b2cf7f, 0x30b5ffe9,
          0xbdbdf21c, 0xcabac28a, 0x53b39330, 0x24b4a3a6,
          0xbad03605, 0xcdd70693, 0x54de5729, 0x23d967bf,
          0xb3667a2e, 0xc4614ab8, 0x5d681b02, 0x2a6f2b94,
          0xb40bbe37, 0xc30c8ea1, 0x5a05df1b, 0x2d02ef8d
        ]


        def self.crc32_hash(key)
          crc = 2 ** 32 - 1
          key.bytes.each do |ch|
            crc = (crc >> 8) ^ CRC32TAB[(crc ^ ch) & 0xff]
          end
          return ((~crc) >> 16) & 0x7fff
        end

      end

    end

  end
end


lib/em-couchbase/error.rb
# Author:: Couchbase <info@couchbase.com>
# Copyright:: 2012 Couchbase, Inc.
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

module EventMachine
  module Protocols

    module Couchbase

      module Error

        # @return [Couchbase::Error::Base]
        def self.map_error_code(code)
          case code
          when 0x00   # PROTOCOL_BINARY_RESPONSE_SUCCESS
            nil
          when 0x01   # PROTOCOL_BINARY_RESPONSE_KEY_ENOENT
            NotFound
          when 0x02   # PROTOCOL_BINARY_RESPONSE_KEY_EEXISTS
            KeyExists
          when 0x03   # PROTOCOL_BINARY_RESPONSE_E2BIG
            TooBig
          when 0x04   # PROTOCOL_BINARY_RESPONSE_EINVAL
            Invalid
          when 0x05   # PROTOCOL_BINARY_RESPONSE_NOT_STORED
            NotStored
          when 0x06   # PROTOCOL_BINARY_RESPONSE_DELTA_BADVAL
            DeltaBadval
          when 0x07   # PROTOCOL_BINARY_RESPONSE_NOT_MY_VBUCKET
            NotMyVbucket
          when 0x20   # PROTOCOL_BINARY_RESPONSE_AUTH_ERROR
            Auth
          when 0x22   # PROTOCOL_BINARY_RESPONSE_ERANGE
            Range
          when 0x81   # PROTOCOL_BINARY_RESPONSE_UNKNOWN_COMMAND
            UnknownCommand
          when 0x82   # PROTOCOL_BINARY_RESPONSE_ENOMEM
            NoMemory
          when 0x83   # PROTOCOL_BINARY_RESPONSE_NOT_SUPPORTED
            NotSupported
          when 0x84   # PROTOCOL_BINARY_RESPONSE_EINTERNAL
            Internal
          when 0x85   # PROTOCOL_BINARY_RESPONSE_EBUSY
            Busy
          else
            Base
          end
        end

        class Base < ::RuntimeError
          # @return [Fixnum] the error code from libcouchbase
          attr_accessor :error
          # @return [String] the key which generated error */
          attr_accessor :key
          # @return [Fixnum] the version of the key (+nil+ unless accessible)
          attr_accessor :cas
          # @return [Symbol] the operation (+nil+ unless accessible)
          attr_accessor :operation
        end

        # Authentication error
        class Auth < Base; end

        # The given bucket not found in the cluster */
        class BucketNotFound < Base; end

        # The cluster is too busy now. Try again later */
        class Busy < Base; end

        # The given value is not a number */
        class DeltaBadval < Base; end

        # Internal error */
        class Internal < Base; end

        # Invalid arguments */
        class Invalid < Base; end

        # Key already exists */
        class KeyExists < Base; end

        # Network error */
        class Network < Base; end

        # Out of memory error */
        class NoMemory < Base; end

        # No such key */
        class NotFound < Base; end

        # The vbucket is not located on this server */
        class NotMyVbucket < Base; end

        # Not stored */
        class NotStored < Base; end

        # Not supported */
        class NotSupported < Base; end

        # Invalid range */
        class Range < Base; end

        # Temporary failure. Try again later */
        class TemporaryFail < Base; end

        # Object too big */
        class TooBig < Base; end

        # Unknown command */
        class UnknownCommand < Base; end

        # Unknown host */
        class UnknownHost < Base; end

        # Failed to decode or encode value */
        class ValueFormat < Base; end

        # Protocol error */
        class Protocol < Base; end

        # Timeout error */
        class Timeout < Base; end

        # Connect error */
        class Connect < Base; end
      end

    end
  end
end


lib/em-couchbase/packet.rb
# Author:: Couchbase <info@couchbase.com>
# Copyright:: 2012 Couchbase, Inc.
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

module EventMachine
  module Protocols

    module Couchbase

      module Packet

        #   uint8_t   magic
        #   uint8_t   opcode
        #   uint16_t  keylen
        #   uint8_t   extlen
        #   uint8_t   datatype
        #   uint16_t  vbucket
        #   uint32_t  bodylen
        #   uint32_t  opaque
        #   uint64_t  cas
        REQUEST_HEADER_FMT = "CCnCCnNNQ"
        REQUEST_HEADER_SIZE = 24  # bytes

        # Sum of lengths of fields below
        #   uint8_t   magic
        #   uint8_t   opcode
        #   uint16_t  keylen
        #   uint8_t   extlen
        #   uint8_t   datatype
        #   uint16_t  status
        #   uint32_t  bodylen
        #   uint32_t  opaque
        #   uint64_t  cas
        RESPONSE_HEADER_FMT = "CCnCCnNNQ"
        RESPONSE_HEADER_SIZE = 24 # bytes

        CMD_GET = 0x00
        CMD_SET = 0x01
        CMD_ADD = 0x02
        CMD_REPLACE = 0x03
        CMD_DELETE = 0x04
        CMD_INCREMENT = 0x05
        CMD_DECREMENT = 0x06
        CMD_QUIT = 0x07
        CMD_FLUSH = 0x08
        CMD_GETQ = 0x09
        CMD_NOOP = 0x0a
        CMD_VERSION = 0x0b
        CMD_GETK = 0x0c
        CMD_GETKQ = 0x0d
        CMD_APPEND = 0x0e
        CMD_PREPEND = 0x0f
        CMD_STAT = 0x10
        CMD_SETQ = 0x11
        CMD_ADDQ = 0x12
        CMD_REPLACEQ = 0x13
        CMD_DELETEQ = 0x14
        CMD_INCREMENTQ = 0x15
        CMD_DECREMENTQ = 0x16
        CMD_QUITQ = 0x17
        CMD_FLUSHQ = 0x18
        CMD_APPENDQ = 0x19
        CMD_PREPENDQ = 0x1a
        CMD_VERBOSITY = 0x1b
        CMD_TOUCH = 0x1c
        CMD_GAT = 0x1d
        CMD_GATQ = 0x1e
        CMD_SASL_LIST_MECHS = 0x20
        CMD_SASL_AUTH = 0x21
        CMD_SASL_STEP = 0x22

        def self.build(opaque, vbucket, opcode, *args)
          case opcode
          when :set
            key, value, flags, expiration, cas = args.shift(5)
            bodylen = key.size + value.size + 8
            [
              0x80,             # uint8_t   magic
              CMD_SET,          # uint8_t   opcode
              key.size,         # uint16_t  keylen
              8,                # uint8_t   extlen (flags + expiration)
              0,                # uint8_t   datatype
              vbucket,          # uint16_t  vbucket
              bodylen,          # uint32_t  bodylen
              opaque || 0,      # uint32_t  opaque
              cas || 0,         # uint64_t  cas
              flags || 0,       # uint32_t  flags
              expiration || 0,  # uint32_t  expiration
              key,
              value
            ].pack("#{REQUEST_HEADER_FMT}NNa*a*")
          when :get
            key = args.shift
            bodylen = key.size
            [
              0x80,             # uint8_t   magic
              CMD_GET,          # uint8_t   opcode
              key.size,         # uint16_t  keylen
              0,                # uint8_t   extlen
              0,                # uint8_t   datatype
              vbucket,          # uint16_t  vbucket
              bodylen,          # uint32_t  bodylen
              opaque || 0,      # uint32_t  opaque
              0,                # uint64_t  cas
              key
            ].pack("#{REQUEST_HEADER_FMT}a*")
          when :incr, :decr
            cmd_id = opcode == :incr ? CMD_INCREMENT : CMD_DECREMENT
            key, delta, initial, expiration, cas = args.shift(5)
            delta ||= 1
            initial ||= 0
            bodylen = key.size + 20
            [
              0x80,                   # uint8_t   magic
              cmd_id,                 # uint8_t   opcode
              key.size,               # uint16_t  keylen
              20,                     # uint8_t   extlen (delta + initial + expiration)
              0,                      # uint8_t   datatype
              vbucket,                # uint16_t  vbucket
              bodylen,                # uint32_t  bodylen
              opaque || 0,            # uint32_t  opaque
              cas || 0,               # uint64_t  cas
              delta >> 32,            # uint64_t
              delta & 0xffffffff,     #
              initial >> 32,          # uint64_t
              initial & 0xffffffff,   #
              expiration || 0,        # uint32_t
              key
            ].pack("#{REQUEST_HEADER_FMT}NNNNNa*")
          when :sasl_auth
            mech, username, password = args.shift(3)
            value = "\0#{username}\0#{password}"
            bodylen = mech.size + value.size
            [
              0x80,             # uint8_t   magic
              CMD_SASL_AUTH,    # uint8_t   opcode
              mech.size,        # uint16_t  keylen
              0,                # uint8_t   extlen
              0,                # uint8_t   datatype
              0,                # uint16_t  vbucket
              bodylen,          # uint32_t  bodylen
              0,                # uint32_t  opaque
              0,                # uint64_t  cas
              mech,
              value
            ].pack("#{REQUEST_HEADER_FMT}a*a*")
          else
            raise Couchbase::Error::UnknownCommand, [opcode, *args].inspect
          end
        end

        def self.parse(data)
          while data.length >= RESPONSE_HEADER_SIZE
            header = data[0...RESPONSE_HEADER_SIZE]
            ( magic,
              opcode,
              keylen,
              extlen,
              datatype,
              status,
              bodylen,
              opaque,
              cas ) = header.unpack(RESPONSE_HEADER_FMT)

            if magic != 0x81
              fail Couchbase::Error::Protocol.new "Broken packet: #{header.inspect}"
            end

            if data.size < bodylen + RESPONSE_HEADER_SIZE
              return  # need moar data
            else
              data[0...RESPONSE_HEADER_SIZE] = ""
            end

            ext = data[0...extlen]
            data[0...extlen] = ""

            key = data[0...keylen]
            data[0...keylen] = ""

            vallen = bodylen - extlen - keylen
            body = data[0...vallen]
            data[0...vallen] = ""

            result = Result.new(:key => key,
                                :value => body,
                                :status => status,
                                :cas => cas)

            case opcode
            when CMD_SET
              result.operation = :set
            when CMD_SASL_AUTH
              result.operation = :sasl_auth
            when CMD_GET
              result.operation = :get
              result.flags, _ = ext.unpack("N")
            when CMD_INCREMENT
              result.operation = :incr
              hi, lo = result.value.unpack("NN")
              result.value = hi << 32 | lo
            when CMD_DECREMENT
              result.operation = :decr
              hi, lo = result.value.unpack("NN")
              result.value = hi << 32 | lo
            else
              raise Couchbase::Error::UnknownCommand, header.inspect
            end

            if error_class = Couchbase::Error.map_error_code(status)
              result.error = error_class.new(body)
              result.error.error = status
              result.error.key = key
              result.error.cas = cas
              result.error.operation = result.operation
            end
            yield(result.operation, opaque, result)
          end
        end
      end

    end

  end
end
